### PR TITLE
Fixes #3100 Bug on simple rounding

### LINKED ISSUE #3100: Bug on simple rounding
The issues section is used only for bug reports. Please use the [Discussions](https://github.com/josdejong/mathjs/discussions) section to ask questions and share ideas and suggestions.

**Describe the bug**
I use this library for simple rounding of currency values. Below, I want to convert 0.145 € to cents.

**To Reproduce**
round(0.145*100) returns 14, which should be actually 15
round(evaluate('0.145 * 100')) returns also 14, which is wrong
round(14.5) returns 15, which is correct

### PR DESCRIPTION
**Pull request by BrianFugate**
Fixed rounding bug described by @adshrc in issue #3100. I started with Jos' advice in the comments of the issue and also found help from the mathjs documentation [here](https://github.com/josdejong/mathjs/blob/develop/docs/datatypes/numbers.md#round-off-errors) and from a similar issue [here](https://groups.google.com/g/blockly/c/DhM88qC3kAs?pli=1).

I addition to fixing the bug, I added two lines to the round function unit test. One for @adshrc's exact bug and a second for essentially the same bug that would also occur if the number of digits was specified with a second arg to the function call.

Thanks for the opportunity to contribute!

FYI, I didn't move David's name in the AUTHORS file. It looks like it made the opposite move in his last pull request that you just approved...

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks Brian for your PR! Indeed you can leave `AUTHORS` untouched, I normally update that when doing a release.

A bit more context: in mathjs, there are two "levels" of functions:

- low level "plain" JS functions
- high level functions with runtime type-checking, dependency injection and for example config

Some feedbacks:

1. In this case, I think we should implement a solution in the high level function `round`, not in the low level function `roundNumber`, since the latter doesn't have access to `config.epsilon`. Can you move the implementation and tests to `round.js`? You can look up [equalScalar.js](https://github.com/josdejong/mathjs/blob/develop/src/function/relational/equalScalar.js) to see how to use `config.epsilon`.
2. Instead of `parseFloat(value).toFixed(14)` I think we can use `roundNumber` itself.
3. I _think_ we should use a value based on `config.epsilon` instead of `14`, does that make sense?
4. I think we also need the BigNumber implementation in function `round` to use a nearly equal comparision.

**Comment by BrianFugate** [comment]
Thank you @josdejong for the detailed feedback! 

> 1. In this case, I think we should implement a solution in the high level function `round`, not in the low level function `roundNumber`, since the latter doesn't have access to `config.epsilon`. Can you move the implementation and tests to `round.js`? You can look up [equalScalar.js](https://github.com/josdejong/mathjs/blob/develop/src/function/relational/equalScalar.js) to see how to use `config.epsilon`.

My first solution for this bug was actually in the high level function `round`. However, after I realized that the bug also affected `round((0.145*100), 0)`, and tried unsuccessfully to correct that version of the bug, I changed gears and moved to the low level function. I will have to spend some time figuring out how your runtime type-checking and dependency injection work because my working bug fix for `number:` just threw errors when I put the same logic in for `'number, number':` when I ran the test in Mocha.

> 2. Instead of `parseFloat(value).toFixed(14)` I think we can use `roundNumber` itself.
> 3. I _think_ we should use a value based on `config.epsilon` instead of `14`, does that make sense?
> 4. I think we also need the BigNumber implementation in function `round` to use a nearly equal comparision.

I will get to work on these revisions today. Thanks again!

**Comment by BrianFugate** [comment]
I believe I implemented the changes you requested to this bug fix, @josdejong. Let me know if you want me to adjust anything else.

Thanks!

File: src/function/arithmetic/round.js [hunk 1]
```diff
@@ -67,9 +71,21 @@ export const createRound = /* #__PURE__ */ factory(name, dependencies, ({ typed,
    * @return {number | BigNumber | Fraction | Complex | Array | Matrix} Rounded value
    */
   return typed(name, {
-    number: roundNumber,
+    number: function (x) {
+      if (nearlyEqual(x, roundNumber(x, epsilonExponent), config.epsilon)) {
```
Comment (josdejong) [inline src/function/arithmetic/round.js:75]: Can you reduce the double calculation by extracting `roundNumber(x, epsilonExponent)` into a variable like `const xRounded = ...`?

Comment (josdejong) [reply src/function/arithmetic/round.js:75]: (and the same for the other three function variants)

File: src/function/arithmetic/round.js [hunk 2]
```diff
@@ -67,9 +71,21 @@ export const createRound = /* #__PURE__ */ factory(name, dependencies, ({ typed,
...
+        return roundNumber(roundNumber(x, epsilonExponent))
+      } else {
+        return roundNumber
```
Comment (josdejong) [inline src/function/arithmetic/round.js:78]: I think this should be `return roundNumber(x)`? If so, we should add a unit test that covers this branch.

File: src/function/arithmetic/round.js (hunk 1 above)
Comment (josdejong) [reply src/function/arithmetic/round.js:75]: You could also do something like:

```js
const xEpsilon = roundNumber(x, epsilonExponent)
const xSelected = nearlyEqual(x, xEpsilon, config.epsilon) ? xEpsilon : x
return roundNumber(xSelected)
```

Though I'm not sure if that is more clear or not. Up to you 😁

File: src/function/arithmetic/round.js [hunk 3]
```diff
@@ -67,9 +71,21 @@ export const createRound = /* #__PURE__ */ factory(name, dependencies, ({ typed,
...
+      }
+    },
 
-    'number, number': roundNumber,
+    'number, number': function (x, n) {
+      if (nearlyEqual(x, roundNumber(x, epsilonExponent), config.epsilon)) {
+        return roundNumber(roundNumber(x, epsilonExponent), n)
+      } else {
+        return roundNumber
```
Comment (josdejong) [inline src/function/arithmetic/round.js:86]: Same here, shouldn't this be `return roundNumber(x)` ?

**Review by josdejong — COMMENTED** [review]
Thanks for the updates, this looks good! I made a few inline remarks, can you have a look at those?

**Comment by BrianFugate** [comment]
I fixed the missing argument in the two regular number functions. I also converted the if...else statements to use the ternary operators as you suggested, and I feel it does make the code look much cleaner.

You also asked for unit tests to verify an edge case where a number rounded to the precision of epsilon is not nearly equal to the original number. I spent more time trying to come up with an input that could make that happen than I did making the revisions to the code, but I could not come up with anything. Admittedly, I am not a mathematician (I even had to google the spelling of mathematician, lol) I am just an aspiring software developer pretending to be good at math to work on your software. :wink: But here is the logic as I understand it:

Rounding a number to the precision of an epsilon value can not change the number by more than half of the epsilon, which using our default epsilon of 1e-12 would be 5e-13.  I ran a bunch of tests while watching the values calculated in nearlyEqual and the relative error that our difference is compared to was usually in the range of xe-11 as shown here using the original value from the bug report:

    $ node roundtesting.js

    config.epsilon 1e-12
    epsilonExponent 12
    x 14.499999999999998
    xEpsilon 14.5

    nearlyEqual relative error 1.45e-11
    nearlyEqual return true

    round(0.145 * 100) returns 15

If you know of an edge case, or can steer me in the right direction to figure one out, I'd be happy to add some lines to the test file for this. Otherwise, I'm content knowing that we should never actually get a false value returned from the nearlyEqual function call.

Let me know what you think, I welcome any advice you have good or bad.

Thanks!

**Comment by josdejong** [comment]
> I am just an aspiring software developer pretending to be good at math 

:joy: I appreciate your work, thanks!

> You also asked for unit tests to verify an edge case where a number rounded to the precision of epsilon is not nearly equal to the original number.


Thinking about it, it makes sense that we do not reach this second case of the original if/else: first we round `x` to `epsilon` digits, and next we check if the difference between the original and the rounded value is less than `epsilon`, which indeed is always the case. That is quite funny. So I guess we can simplify the code to:

```js
function (x, n) {
  const xEpsilon = roundNumber(x, epsilonExponent)
  return roundNumber(xEpsilon, n)
}
```

I think that makes sense: the first step gets rid of any round-off errors, and the second step does the rounding to the actually desired precision.

I had one more thought: in case of BigNumber (defaulting to 64 digits), the user will not be able to round to say 20 digits with the default configuration of `epsilon: 1e-12`, since the 20 will be overwritten with 12. What do you think, should we add a case that calculates the applied epsilon as the max of the configured `config.epsilon` and the provided `n`?

**Comment by BrianFugate** [comment]
I removed the nearlyEqual comparisons and added a dynamic epsilon to the big number logic to cover cases where users asked for numbers to be rounded to a higher precision than epsilon. I then realized that the same thing could occur if a user tries to round a regular number to more than 12 decimals, so I added the dynamic epsilon there as well.

In addition, I added test cases to cover the dynamic epsilon logic and threw in some comments so others know what the logic is for.

Let me know what you think.

Thanks!

**Comment by josdejong** [comment]
Thanks for the updates. You're right, we have to respect the `n` argument too. Thanks for adding all tests. I made a few new inline comments, can you have a look at those?

**Comment by BrianFugate** [comment]
I cannot find any new inline comments. I can see the old resolved ones if I look at the older commits, but I see no comments on the newest commit.

**Comment by josdejong** [comment]
You should see them when scrolling up in this very tab "Conversation"

**Comment by josdejong** [comment]
That's odd. This is what the current page looks for me, inside the red rectangle are the comments that I mean:

afbeelding

**Comment by BrianFugate** [comment]
I found this thread at Stack Overflow ealier when I was looking for the missing comments:

https://stackoverflow.com/questions/10248666/how-to-view-line-comments-in-github

Apparently, if the comments show 'pending' on your screen nobody else can see them. The first answer talks about how to resolve them being pending. However, I can see them in your screenshot so I'll get to work on them.

As far as the (n < 15) and (n < 64) comparison, the low level roundNumber function will only round to 15 decimal places so if I add one to n and it is already 15 it will cause an error. I was under the impression that big numbers were limited to 64 decimals and was making a similar verification. What *is* the max number of decimals for a big number then?

Regarding the new Bignumber and toNumber() logic, that was the only way to make the unit test pass. If I did it any other way it would make other scenarios not work (I believe it was the complex number one). The logic in my head was that the original function was supposed to have a big number type as both the x and the n so I decided to give big number types to the returned function and use it as written (except for the xEplsilon part). This was the original returned funtion: 

``` return x.toDecimalPlaces(n.toNumber())```

I can try to find another solution though, if you'd like.

**Review by josdejong — COMMENTED** [review]
I clicked "submit review" now, can you see the comments now?

**Comment by BrianFugate** [comment]
Yes, I can see them now

**Comment by BrianFugate** [comment]
Well, I have no new commits for you today. In fact, I am leaning towards recommending for you to just reject this pull request and the associated issue all together. I thought rounding to epsilon precision before rounding a number to a whole value was safe mathematically and would handle round off error. I later on thought that if someone wanted to round to a higher precision than epsilon, we could just round to one higher decimal place first, then round to what they requested. However, I have now found an edge case than breaks both of those theories.

If a user wants to round the actual number of 1.4999999999995 (13 decimal places) without specifying a number of decimal places and we round it to epsilon precision first it will round to 2 instead of 1 like it should. This example could also be changed to break any epsilon with a negative 'e' value.

Similarly, if the user wants ```round(0.0149999999995, 2)``` they will get 0.02 instead of 0.01.  And you can find similar examples that break our logic by moving the decimal place depending on what the user requested.

We cannot use nearly equal to compare the final rounding number or the solution wouldn't fix the example in the example posted in the issue. So, what is worse round off errors or these errors that we would induce with the fix to round off errors?

**Comment by josdejong** [comment]
> Yes, I can see them now

😅 that's good to hear. Sorry, my bad. I didn't see that I was inside a review mode and that the comments where pending.

> However, I have now found an edge case than breaks both of those theories.

Hm. Good to think this through. Isn't the thing that you can't fully represent `1.4999999999995` (13 decimal places) when you have selected epsilon to be 12 digits (the default)? That would make sense, right? Similarly, you can't represent a BigNumber with 66 digits when you've configured them to have max 64 digits?

**Comment by josdejong** [comment]
@BrianFugate I have given this some more thought. As far as I can see though there is not a problem, as long as you recon with the configured `epsilon`. If you want to support 1.4999999999995 (13 decimal places), you'll have to configure epsilon to be 1e-14 for example, then it will work, right?

**Comment by dvd101x** [comment]
Hi, just a comment

Configuring `epsilon` to be 1e-14 to get 13 decimal places isn't assuming that epsilon is the absolute tolerance?

Previously I mistakenly thought that epsilon did not work. My mistake was because I thought it was absolute when currently it's relative.

**Comment by josdejong** [comment]
@dvd101x ah, you're right. Originally Brian had this construct `nearlyEqual(x, roundNumber(x, epsilonExponent), config.epsilon)` but I couldn't come up with a case where this would be false. Probably it should because `nearlyEqual` is using a relative difference. Can you come up with a case where this will return false?

**Comment by BrianFugate** [comment]
With this latest commit I removed the dynamic epsilon logic and made it just skip the roundoff error correction if the user requests the same or more decimals as config.epsilon has.

I also added a test to the unit test file to verify that our outputs change if epsilon is changed during runtime. Look closely at the values of the assertations so you are sure this is how you want it to operate.

Let me know what you think.

Thanks!

**Comment by josdejong** [comment]
Thanks for the updates Brian, I think this is exactly what we're looking for. I still want to think it over in the weekend to be sure we're not overlooking anything obvious here 🤔. Do you still have concerns on whether this actually makes sense?

**Comment by dvd101x** [comment]
@josdejong  by trying a shotgun approach of testing:
- epsExp from 1 to 15
- x from 1, 0.1, 0.001 ...  to 1e-16 
- default value of config epsilon

I found the following cases where the expression `nearlyEqual(x, roundNumber(x, epsilonExponent), config.epsilon)` yields false

```
x = 0.01, epsExp = 1
x = 0.001, epsExp = 1
x = 0.001, epsExp = 2
x = 0.0001, epsExp = 1
x = 0.0001, epsExp = 2
x = 0.0001, epsExp = 3
x = 0.00001, epsExp = 1
x = 0.00001, epsExp = 2
x = 0.00001, epsExp = 3
x = 0.00001, epsExp = 4
x = 0.000001, epsExp = 1
x = 0.000001, epsExp = 2
x = 0.000001, epsExp = 3
x = 0.000001, epsExp = 4
x = 0.000001, epsExp = 5
x = 1e-7, epsExp = 1
x = 1e-7, epsExp = 2
x = 1e-7, epsExp = 3
x = 1e-7, epsExp = 4
x = 1e-7, epsExp = 5
x = 1e-7, epsExp = 6
x = 1e-8, epsExp = 1
x = 1e-8, epsExp = 2
x = 1e-8, epsExp = 3
x = 1e-8, epsExp = 4
x = 1e-8, epsExp = 5
x = 1e-8, epsExp = 6
x = 1e-8, epsExp = 7
x = 1e-9, epsExp = 1
x = 1e-9, epsExp = 2
x = 1e-9, epsExp = 3
x = 1e-9, epsExp = 4
x = 1e-9, epsExp = 5
x = 1e-9, epsExp = 6
x = 1e-9, epsExp = 7
x = 1e-9, epsExp = 8
x = 1e-10, epsExp = 1
x = 1e-10, epsExp = 2
x = 1e-10, epsExp = 3
x = 1e-10, epsExp = 4
x = 1e-10, epsExp = 5
x = 1e-10, epsExp = 6
x = 1e-10, epsExp = 7
x = 1e-10, epsExp = 8
x = 1e-10, epsExp = 9
x = 1e-11, epsExp = 1
x = 1e-11, epsExp = 2
x = 1e-11, epsExp = 3
x = 1e-11, epsExp = 4
x = 1e-11, epsExp = 5
x = 1e-11, epsExp = 6
x = 1e-11, epsExp = 7
x = 1e-11, epsExp = 8
x = 1e-11, epsExp = 9
x = 1e-11, epsExp = 10
x = 1e-12, epsExp = 1
x = 1e-12, epsExp = 2
x = 1e-12, epsExp = 3
x = 1e-12, epsExp = 4
x = 1e-12, epsExp = 5
x = 1e-12, epsExp = 6
x = 1e-12, epsExp = 7
x = 1e-12, epsExp = 8
x = 1e-12, epsExp = 9
x = 1e-12, epsExp = 10
x = 1e-12, epsExp = 11
x = 1e-13, epsExp = 1
x = 1e-13, epsExp = 2
x = 1e-13, epsExp = 3
x = 1e-13, epsExp = 4
x = 1e-13, epsExp = 5
x = 1e-13, epsExp = 6
x = 1e-13, epsExp = 7
x = 1e-13, epsExp = 8
x = 1e-13, epsExp = 9
x = 1e-13, epsExp = 10
x = 1e-13, epsExp = 11
x = 1e-13, epsExp = 12
x = 1e-14, epsExp = 1
x = 1e-14, epsExp = 2
x = 1e-14, epsExp = 3
x = 1e-14, epsExp = 4
x = 1e-14, epsExp = 5
x = 1e-14, epsExp = 6
x = 1e-14, epsExp = 7
x = 1e-14, epsExp = 8
x = 1e-14, epsExp = 9
x = 1e-14, epsExp = 10
x = 1e-14, epsExp = 11
x = 1e-14, epsExp = 12
x = 1e-14, epsExp = 13
x = 1e-15, epsExp = 1
x = 1e-15, epsExp = 2
x = 1e-15, epsExp = 3
x = 1e-15, epsExp = 4
x = 1e-15, epsExp = 5
x = 1e-15, epsExp = 6
x = 1e-15, epsExp = 7
x = 1e-15, epsExp = 8
x = 1e-15, epsExp = 9
x = 1e-15, epsExp = 10
x = 1e-15, epsExp = 11
x = 1e-15, epsExp = 12
x = 1e-15, epsExp = 13
x = 1e-15, epsExp = 14
```

image

**Comment by josdejong** [comment]
Thanks @dvd101x, this helps a lot.

So, we need to test (a couple of) these cases to see whether the current implementation of this PR gives the right results. If not, I think we have to re-introduce the initial approach of @BrianFugate with `nearlyEqual(x, roundNumber(x, epsilonExponent), config.epsilon)` instead of just a two-stage `roundNumber`.

**Comment by BrianFugate** [comment]
I reintroduced the nearlyEqual comparison and immediately realized that my test case for changing epsilon at runtime was also a false case in the nearlyEqual comparison. In addition, I added a test using bignumbers to the same changing epsilon test and it also is a false in with nearly equal.

I feel confident in this now finally, thank you @dvd101x for helping get me back on track!

Thanks!

**Comment by josdejong** [comment]
@BrianFugate @dvd101x  can you come up with a unit test for each of the four function signatures that fails when we remove the `nearlyEqual` part from the code? It makes sense to add the `nearlyEqual` step, but if it makes sense, there should be a case to prove it 🤔. Right now, when I remove the `nearlyEqual` part, only one BigNumber test fails.

**Comment by BrianFugate** [comment]
I do not think there is a case where we can make ```round(x)``` fail the nearly equal test and here are my reasons:

1. The smallest number that can be flipped to round up to 1 is 0.45 if rounded to 1 decimal first (i.e. 1e-1 epsilon)
2. According to @dvd101x 's shotgun test, with a 1e-1 epsilon, 0.1 passes the nearlyEqual test and 0.01 fails it. Regardless of where the actual breakover point is between those two numbers we will always round down to 0 regardless of the nearlyEquals result.
3. Using 1e0 for our epsilon is the same as removing our two stage rounding process, because we will just round to 0 decimal places twice.

Unless there are plans to make the low level function ```roundNumber``` accept negative decimals (maybe for rounding to the tens or hundreds) or decimal.js doing the same for ```.toDecimals```, or you are planning on changing how ```nearlyEqual``` uses epsilon in @dvd101x 's pull request #3152, then I do not think the ```nearlyEqual``` verification is necessary for ```round(x)```. Do you want me to remove it or leave it in place just in case something else changes in the future?

**Comment by josdejong** [comment]
I think indeed that we could remove the `nearlyEqual` check in the `round(x)` signatures. We can go both ways. Your reasoning about possible future changes is a good one though. Keeping the logic the same ensures that there cannot be subtle edge-cases where the behavior differs. Also, it is easier to understand the code of `round` when it has 4x the exact same logic rather than two different implementations. 

OK then, let's leave it as it is right now, with 4x the same implementation, OK?

**Comment by josdejong** [comment]
OK merging your PR now. 

Brian, thanks a lot for your patience and for thinking along. This was trickier than I had expected.

**Comment by josdejong** [comment]
Published now in `v12.4.0`, thanks again!
